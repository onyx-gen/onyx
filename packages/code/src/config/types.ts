import type { Mode, Unit } from '@onyx/types'
import type { Theme } from '@unocss/preset-mini'

export interface Configuration {
  mode: Mode
  unit: Unit
  variantGroup: boolean
  nearestInference: boolean
  theme: Theme
}
