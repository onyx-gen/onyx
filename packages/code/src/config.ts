import type { Mode } from '@onyx/types'

interface OnyxConfiguration {
  mode: Mode
}

const config: OnyxConfiguration = {
  mode: 'variables',
}

export default config
