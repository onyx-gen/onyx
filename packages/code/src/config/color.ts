import type { Theme } from '@unocss/preset-mini'
import type { HEX } from '@onyx-gen/types'

export type InferenceColorMap = Record<HEX, string[]>

function isHEX(value: string): boolean {
  return /^#[0-9A-F]{6}$/i.test(value)
}

/**
 * Maps colors to their corresponding names.
 *
 * @param {NonNullable<Theme['colors]>} colors - The colors object to be mapped.
 * @param {string} prefix - The prefix to append color name.
 * @return {InferenceColorMap} - The resulting object that maps HEX colors to their corresponding names.
 */
export function createColorLookup(
  colors: NonNullable<Theme['colors']>,
  prefix: string = '',
): InferenceColorMap {
  const result: InferenceColorMap = {}

  for (const colorName in colors) {
    const value = colors[colorName]
    if (typeof value === 'string') { // leaf node
      if (isHEX(value)) {
        if (!result[value])
          result[value] = []

        result[value].push(prefix + colorName)
      }
    }
    else { // object node
      const nestedResults = createColorLookup(value, `${prefix}${colorName}-`)
      Object.assign(result, nestedResults)
    }
  }

  return result
}
