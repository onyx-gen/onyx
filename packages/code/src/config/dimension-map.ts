export type InferenceDimensionMap = Record<number, string>

export const dimensionMap: InferenceDimensionMap = {
  0: '0',
  1: 'px',
  2: '0.5',
  4: '1',
  6: '1.5',
  8: '2',
  10: '2.5',
  12: '3',
  14: '3.5',
  16: '4',
  20: '5',
  24: '6',
  28: '7',
  32: '8',
  36: '9',
  40: '10',
  44: '11',
  48: '12',
  56: '14',
  64: '16',
  80: '20',
  96: '24',
  112: '28',
  128: '32',
  144: '36',
  160: '40',
  176: '44',
  192: '48',
  208: '52',
  224: '56',
  240: '60',
  256: '64',
  288: '72',
  320: '80',
  384: '96',
}
