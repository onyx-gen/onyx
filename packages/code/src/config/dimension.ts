import { h } from '@unocss/preset-mini/utils'
import { round } from '../utils'

export type InferenceDimensionMap = Record<number, string[]>

const values: number[] = [
  0,
  0.5,
  1,
  1.5,
  2,
  2.5,
  3,
  3.5,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  14,
  16,
  20,
  24,
  28,
  32,
  36,
  40,
  44,
  48,
  52,
  56,
  60,
  64,
  72,
  80,
  96,
]

function remToUnitless(r: string): number {
  return Number.parseFloat(round(Number.parseFloat(r.replace('rem', ''))))
}

function createComputedDimensions() {
  const result: Record<string, string> = {}

  for (const value of values.map(String)) {
    const remValue = h.rem(value)
    if (remValue)
      result[value] = remValue
  }

  return result
}

export function createDimensionLookupPx(): InferenceDimensionMap {
  // iterate from 0 to 10
  const result: Record<number, string[]> = {}

  for (let i = 0; i <= 10; i++)
    result[i] = [`${i}`]

  return result
}

export function createDimensionLookup(
  _dimensions: Record<string, string>,
): InferenceDimensionMap {
  const result: Record<number, string[]> = {}

  const filteredRemDimensions = Object.fromEntries(
    Object.entries(_dimensions)
      .filter(([key, value]) => value.endsWith('rem') && key !== 'DEFAULT'),
  )

  function addToResults(d: Record<string, string>) {
    Object.entries(d).forEach(([key, value]) => {
      const unitlessValue = remToUnitless(value)

      if (!result[unitlessValue])
        result[unitlessValue] = []

      result[unitlessValue].push(key)
    })
  }

  addToResults(filteredRemDimensions)
  addToResults(createComputedDimensions())
  addToResults({
    px: '0.0625rem',
  })

  return result
}
