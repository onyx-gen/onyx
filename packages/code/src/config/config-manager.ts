import type { IConfiguration } from '@onyx-gen/types'
import { Configuration } from './config'

class ConfigurationManager {
  private _config: Configuration | null = null

  public get config(): Configuration {
    if (!this._config)
      throw new Error('Configuration has not been loaded yet')
    return this._config
  }

  public async loadConfig(): Promise<Configuration> {
    const savedConfig: IConfiguration | undefined = await figma.clientStorage.getAsync('config')
    this._config = new Configuration(savedConfig)
    return this.config
  }

  public async updateConfig(newConfig: Partial<IConfiguration>): Promise<Configuration> {
    this._config = new Configuration(newConfig, this.config.config)
    await figma.clientStorage.setAsync('config', this.config.config)
    return this.config
  }
}

export default ConfigurationManager
