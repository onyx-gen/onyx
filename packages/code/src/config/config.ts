import { defu } from 'defu'
import type { IConfiguration } from '@onyx/types'
import { theme } from '@unocss/preset-mini'
import type { InferenceColorMap } from './color'
import { createColorLookup } from './color'
import type { InferenceDimensionMap } from './dimension'
import { createDimensionLookup } from './dimension'

interface LookupCache {
  color: InferenceColorMap | null
  dimensions: InferenceDimensionMap | null
  borderDimensions: InferenceDimensionMap | null
}

const defaultConfig: IConfiguration = {
  mode: 'variables',
  unit: 'px',
  variantGroup: true,
  nearestInference: true,
  theme,
  newBuilder: true,
}

export class Configuration implements IConfiguration {
  public mode: IConfiguration['mode']
  public unit: IConfiguration['unit']
  public variantGroup: IConfiguration['variantGroup']
  public nearestInference: IConfiguration['nearestInference']
  public theme: IConfiguration['theme']
  public newBuilder: IConfiguration['newBuilder']

  public readonly config: IConfiguration

  private lookupCache: LookupCache = {
    color: null,
    dimensions: null,
    borderDimensions: null,
  }

  constructor(
    config: Partial<IConfiguration> | undefined,
    ..._defaults: (IConfiguration | undefined)[]
  ) {
    const defaults: IConfiguration[] = _defaults.filter(c => !!c) as IConfiguration[]
    if (config)
      this.config = defu(config, ...defaults, defaultConfig) as IConfiguration
    else
      this.config = defu(defaults[0], ...defaults.slice(1), defaultConfig) as IConfiguration

    this.mode = this.config.mode
    this.unit = this.config.unit
    this.variantGroup = this.config.variantGroup
    this.nearestInference = this.config.nearestInference
    this.theme = this.config.theme
    this.newBuilder = this.config.newBuilder
  }

  public get colorLookup(): InferenceColorMap {
    if (this.lookupCache.color)
      return this.lookupCache.color

    const lookup = createColorLookup(this.theme.colors || {})
    this.lookupCache.color = lookup

    return lookup
  }

  public get dimensionsLookup(): InferenceDimensionMap {
    if (this.lookupCache.dimensions)
      return this.lookupCache.dimensions

    const lookup = createDimensionLookup(this.theme.spacing || {})
    this.lookupCache.dimensions = lookup

    return lookup
  }
}
