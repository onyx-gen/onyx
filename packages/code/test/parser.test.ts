import { describe, expect, it } from 'vitest'
import FigmaNodeParser from '../src/parsers/figma-node.parser'
import { Configuration } from '../src/config/config'
import TextNodeFactory from './factories/text-node.factory'
import FrameNodeFactory from './factories/frame-node.factory'

describe('parsers', () => {
  const config = new Configuration({})

  it('should parse a file', async () => {
    const parser = new FigmaNodeParser({ default: 'default' }, config)

    const textNode: TextNode = new TextNodeFactory().setCharacters('Hello, World!').create()

    const frameNode: FrameNode = new FrameNodeFactory().addChild(textNode).create()

    const result = await parser.parse(frameNode)

    expect(result).toMatchSnapshot()
  })
})
