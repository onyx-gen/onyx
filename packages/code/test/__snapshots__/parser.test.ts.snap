// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parsers > should parse a file 1`] = `
{
  "children": [],
  "data": {
    "css": {
      "default-default": {
        "css": [
          Set {
            "bg-light-50",
          },
        ],
      },
    },
    "type": "container",
  },
}
`;
